bin/cheritest/cheritest_signal.c

```diff
@@ -222,3 +222,68 @@ test_signal_sigaltstack_disable(const struct cheri_test *ctp __unused)
 
 	cheritest_success();
 }
+
+#ifdef __CHERI_PURE_CAPABILITY__
+extern int __sys_sigaction(int, const struct sigaction *, struct sigaction *);
+
+static void *handler_returncap;
+
+static void
+returncap_func(int signum __unused)
+{
+	handler_returncap = __builtin_return_address(0);
+}
+
+void
+test_signal_returncap(const struct cheri_test *ctp __unused)
+{
+	struct sigaction sa;
+	uintmax_t v;
+
+	sa.sa_handler = returncap_func;
+	sigemptyset(&sa.sa_mask);
+	sa.sa_flags = 0;
+	/*
+	 * Note: we call __sys_sigaction directly here to avoid any interposing by
+	 * libthr and ensure that our handler is called directly by the kernel,
+	 * since we want to capture the return capability for sigcode it gives to
+	 * userspace.
+	 */
+	if (__sys_sigaction(SIGUSR1, &sa, NULL) != 0)
+		cheritest_failure_errx("sigaction failed: %s",
+		                       strerror(errno));
+
+	handler_returncap = NULL;
+	if (kill(getpid(), SIGUSR1) != 0)
```

Comment:
Maybe use raise() here?

Reply:
We could do, I just copied it from the other tests. Inclined to be consistent and leave that as a future cleanup?